pulp_ansible/app/tasks/collections.py

```diff
@@ -205,3 +134,218 @@ def _update_highest_version(collection_version):
         collection_version.is_highest = True
         last_highest.save()
         collection_version.save()
+
+
+class AnsibleDeclarativeVersion(DeclarativeVersion):
+    """
+    Subclassed Declarative version creates a custom pipeline for RPM sync.
+    """
+
+    def pipeline_stages(self, new_version):
+        """
+        Build a list of stages feeding into the ContentUnitAssociation stage.
+
+        This defines the "architecture" of the entire sync.
+
+        Args:
+            new_version (:class:`~pulpcore.plugin.models.RepositoryVersion`): The
+                new repository version that is going to be built.
+
+        Returns:
+            list: List of :class:`~pulpcore.plugin.stages.Stage` instances
+
+        """
+        pipeline = [
+            self.first_stage,
+            QueryExistingArtifacts(),
+            ArtifactDownloader(),
+            ArtifactSaver(),
+            QueryExistingContents(),
+            AnsibleContentSaver(),
+            RemoteArtifactSaver(),
+            ResolveContentFutures(),
+        ]
+        for dupe_query_dict in self.remove_duplicates:
+            pipeline.append(RemoveDuplicates(new_version, **dupe_query_dict))
+
+        return pipeline
+
+
+class AnsibleFirstStage(Stage):
+    """
+    The first stage of a pulp_ansible sync pipeline.
+    """
+
+    def __init__(self, remote):
+        """
+        The first stage of a pulp_ansible sync pipeline.
+
+        Args:
+            remote (AnsibleRemote): The remote data to be used when syncing
+
+        """
+        super().__init__()
+        self.remote = remote
+
+        # Interpret download policy
+        self.deferred_download = self.remote.policy != Remote.IMMEDIATE
+
+    async def run(self):
+        """
+        Build and emit `DeclarativeContent` from the ansible metadata.
+        """
+        with ProgressBar(message="Parsing Collection Metadata") as pb:
```

Comment:
One thing our users have asked for is for us to set the `total` argument on these as early as we know them. If you can know the total here already then set it. Or you can set it later with:

```
pb.total = xxxx
pb.save()
```

Does this make sense?